src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/HoverHandler.cs

```diff
@@ -0,0 +1,128 @@
+                throw new ArgumentNullException(nameof(requestInvoker));
+            }
+
+            if (documentManager is null)
+            {
+                throw new ArgumentNullException(nameof(documentManager));
+            }
+
+            if (projectionProvider is null)
+            {
+                throw new ArgumentNullException(nameof(projectionProvider));
+            }
+
+            if (documentMappingProvider is null)
+            {
+                throw new ArgumentNullException(nameof(documentMappingProvider));
+            }
+
+            _joinableTaskFactory = joinableTaskContext.Factory;
+            _requestInvoker = requestInvoker;
+            _documentManager = documentManager;
+            _projectionProvider = projectionProvider;
+            _documentMappingProvider = documentMappingProvider;
+        }
+
+        public async Task<Hover> HandleRequestAsync(TextDocumentPositionParams request, ClientCapabilities clientCapabilities, CancellationToken cancellationToken)
+        {
+            await _joinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+
+            if (!_documentManager.TryGetDocument(request.TextDocument.Uri, out var documentSnapshot))
+            {
+                return null;
+            }
+
+            // Switch to a background thread.
+            await TaskScheduler.Default;
+
+            var projectionResult = await _projectionProvider.GetProjectionAsync(documentSnapshot, request.Position, cancellationToken).ConfigureAwait(false);
+            if (projectionResult == null)
+            {
+                return null;
+            }
+
+            var serverKind = projectionResult.LanguageKind == RazorLanguageKind.CSharp ? LanguageServerKind.CSharp : LanguageServerKind.Html;
+
+            if (serverKind != LanguageServerKind.CSharp)
+            {
+                // Hover is only supported in the C# LSP for now.
+                return null;
+            }
+
+            var textDocumentPositionParams = new TextDocumentPositionParams()
+            {
+                Position = projectionResult.Position,
+                TextDocument = new TextDocumentIdentifier()
+                {
+                    Uri = projectionResult.Uri
+                }
+            };
+
+            var result = await _requestInvoker.RequestServerAsync<TextDocumentPositionParams, Hover>(
+                Methods.TextDocumentHoverName,
+                serverKind,
+                textDocumentPositionParams,
+                cancellationToken).ConfigureAwait(false);
+
+            if (result == null || result.Range == null)
+            {
+                return null;
+            }
+
+            var mappingResult = await _documentMappingProvider.MapToDocumentRangeAsync(projectionResult.LanguageKind, request.TextDocument.Uri, result.Range, cancellationToken).ConfigureAwait(false);
+
+            if (mappingResult == null || mappingResult.HostDocumentVersion != documentSnapshot.Version)
```

Comment:
Are there cases where `result` is not null but `mappingResult` is null?
I'm wondering if we get a non null result from the hover request and we can't map it back, is it worth showing the result anyway with a 0 length range.

Reply:
> Are there cases where result is not null but mappingResult is null?
I think if this request were to fail, we could get a null `mappingResult`: https://github.com/dotnet/aspnetcore-tooling/blob/master/src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/DefaultLSPDocumentMappingProvider.cs#L53-L57

> I'm wondering if we get a non null result from the hover request and we can't map it back, is it worth showing the result anyway with a 0 length range.
But we won't know the positioning of where to show the hover in the `.razor` file. It may show up somewhere random on the document.

For instance if `int foo = 1;` is on line 5 of the `.razor` file but on line 10 on the virtual .cs file, then we would be showing the `foo` quick info at line 10. I think just not showing it may be preferred in that case.

Reply:
> I think if this request were to fail, we could get a null mappingResult

Yeah, I was wondering if you had any examples where that will happen where line 105 produces a non-null result and line 113 produces null. 

> But we won't know the positioning of where to show the hover in the .razor file. It may show up somewhere random on the document.
For instance if int foo = 1; is on line 5 of the .razor file but on line 10 on the virtual .cs file, then we would be showing the foo quick info at line 10. I think just not showing it may be preferred in that case.

In those cases, you can just return the range to be `new Range(request.Position, request.Position)`.

Reply:
> Yeah, I was wondering if you had any examples where that will happen where line 105 produces a non-null result and line 113 produces null.

Nothing specific comes to mind. Maybe an extremely unlikely event in which the file is deleted from disk between the two lines?

> In those cases, you can just return the range to be new Range(request.Position, request.Position).

Sure that makes sense, updated :)